File: algorithms/189.Rotate_Array/Rotate_Array.java
#include <bits/stdc++.h>

using namespace std;

class Solution {
private:
    void reverse_array(int nums[], int a, int b) {
        while (a < b) {
            swap(nums[a], nums[b]);
            ++a;
            --b;
        }
    }

public:
    void rotate(int nums[], int n, int k) {
        if (n == 0 || k % n == 0)
            return;
        k = k % n;
        reverse_array(nums, 0, n - k - 1);
        reverse_array(nums, n - k, n - 1);
        reverse_array(nums, 0, n - 1);
    }
};

class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int nowIndex = 0, nextIndex;
        int tmp1, tmp2 = nums[0];
        for(int j=0,i=0; j<nums.size(); j++){
            tmp1 = tmp2;
            nowIndex = (k + nowIndex) % (nums.size());
            tmp2 = nums[nowIndex];
            nums[nowIndex] = tmp1;
            if(nowIndex == i) {
                nowIndex = ++i;
                tmp2 = nums[nowIndex];
            }
        }
    }
};
int main() {
    int num[] = {1, 2, 3, 4, 5};
    Solution s;
    s.rotate(num, 5, 3);
    for (int i = 0; i < 5; ++i)
        cout << num[i] << ' ';
    cout << endl;
    return 0;
}

File: algorithms/189.Rotate_Array/Rotate_Array.py
class Solution:
    # @param nums, a list of integer
    # @param k, num of steps
    # @return nothing, please modify the nums list in-place.
    def rotate(self, nums, k):
        k %= len(nums)
        self.reverse(nums, 0, len(nums))
        self.reverse(nums, 0, k)
        self.reverse(nums, k, len(nums))
    
    def reverse(self, nums, start, end):
        while start < end:
            nums[start], nums[end-1] = nums[end-1], nums[start]
            start += 1
            end -= 1

from fractions import gcd

class Solution2:
    # @param nums, a list of integer
    # @param k, num of steps
    # @return nothing, please modify the nums list in-place.
    def rotate(self, nums, k):
        k %= len(nums)
        num_cycles = gcd(len(nums), k)
        cycle_len = len(nums) / num_cycles
        for i in xrange(num_cycles):
            self.apply_cycle_permutation(k, i, cycle_len, nums)
    
    def apply_cycle_permutation(self, k, offset, cycle_len, nums):
        tmp = nums[offset]
        for i in xrange(1, cycle_len):
            nums[(offset+i*k) % len(nums)], tmp = tmp, nums[(offset+i*k) % len(nums)]
        nums[offset] = tmp
          
if __name__ == '__main__':
    nums = [1,2,3,4,5,6,7]
    Solution().rotate(nums, 3)
    print nums

